### PR TITLE
Feat: Initial implementation of MongodbManager

### PR DESCRIPTION
**Pull request by Tailp**
@danglotb @monperrus hi :) , this is the initial implementation of MongodbManager. Smpt, restful mode, JSAP options will be added later .

### CODE REVIEW

File: dspot/src/main/java/eu/stamp_project/utils/collector/DspotInformationCollector.java [hunk 1]
```diff
@@ -10,4 +10,6 @@
 	void reportSelectorInformation(String info);
 	// collect each test's path
 	void reportAmpTestPath(String pathName);
+	// send info over to some other instance
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/collector/DspotInformationCollector.java]: for method level, comments, use Javadoc comment with `/** ... */`

File: dspot/src/main/java/eu/stamp_project/utils/report/output/selector/coverage/json/TestClassJSON.java [hunk 2]
```diff
@@ -43,9 +43,9 @@ public boolean addTestCase(TestCaseJSON testCaseJSON) {
 
 	public String toString() {
 		JSONObject subJson = new JSONObject()
-                  .put("initialCoverage", this.initialInstructionCovered)
```
Comment (monperrus) [inline dspot/src/main/java/eu/stamp_project/utils/report/output/selector/coverage/json/TestClassJSON.java]: why that change? it seems unrelated.

Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/report/output/selector/coverage/json/TestClassJSON.java]: Hi :) , I wanted to have it already capitalized at first letter so I don't need to capitalize it later. Also since Pitmutant use int so this make the code shorter to also have all the values as int, so I only need to use .getInt() later in MongodbManager.

Comment (monperrus) [reply dspot/src/main/java/eu/stamp_project/utils/report/output/selector/coverage/json/TestClassJSON.java]: Is that required for this PR?

**Comment by danglotb** [comment]
Hello @Tailp,

Your changes look good.

Could we move the field `collector` from `Configuration` to `InputConfiguration`?

Also, I think that It would be better for users if we add a specific command-line option to say which collector we want to use, here a `MongoDBCollector`.

I propose to create:

1. an Enum that do this, you look at the enum for Selector, Amplifier and Budgtizer and do the same trick.
2. creating a method `initialize` inside the interface `DspotInformationCollector`.
3. rename the class `MongodbManager` to `MongodbCollector`

On my command-line, I would have something like:
```
java -jar dspot.jar .... --collector MongoDB
```
or 
```
java -jar dspot.jar .... --collector MongoDB --mongo-url mongodb://localhost:27017
```

It will be easier to implement new collectors, right?

WDYT?

File: dspot/src/main/java/eu/stamp_project/utils/report/output/selector/coverage/json/TestClassJSON.java (hunk 2 above)
Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/report/output/selector/coverage/json/TestClassJSON.java]: Hi :) , I have revert the changes to that part in the latest commit.

**Comment by Tailp** [comment]
@danglotb hi :) , thanks for the suggestion . I more or less did according to your idea in the latest commit . I found out method initialize is not neccessary since we can just create set/get methods to fetch jsap informations directly from InputConfiguration just when we instantiate it making things easier. Indeed the extra option makes things easier :) .

File: dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java [hunk 3]
```diff
@@ -346,8 +349,62 @@ public static void setUp(List<String> amplifiers, String budgetizer,
...
+            executeTestsInParallel, numberParallelExecutionProcessors);
+        collectorInit(mongoUrl,collector);
+    }
+
+
+    /* DspotInformation collector related */
+    private static DspotInformationCollector collector = new NullCollector();
+
+    public static void collectorInit(String mongoUrl, String collector) {
+        InputConfiguration.get().setMongoUrl(mongoUrl).setInformationCollector(CollectorEnum.valueOf(collector).getCollector());
```
Comment (danglotb) [inline dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java]: Hello @Tailp 
here, the call to `setMongoUrl` makes this statement mongo specific. Could we have a global API shared by all the `Collector`s? Do we need to keep the mongoUrl inside the `InputConfiguration` or could we put it inside the `MongoCollector`?

**Comment by danglotb** [comment]
Unless my small comment, I'm okay with the current changes. Thank @Tailp very good work!

File: dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java (hunk 3 above)
Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java]: Hi @danglotb :) , I used InputConfiguration as a global config for storing my collector parsed JSAP values , so I can just access those with a getter later . This also make things easier later on for different kinds of collector since it can just fetch the stored JSAP values from inputConfiguration upon instantiation without any initialization :).

Comment (danglotb) [reply dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java]: Yes, I agree with you, However, I'm asking myself: Is anywhere else than in the `mongoCollector` where you use the `mongoUrl`? If no, It seems legit to put this url inside the `mongoCollector`, right? As field for instance, WDYT?

Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java]: @danglotb sure :) , but I might have to add an initialize method to the interface as a result to keep the polymorphism Informationcollector.

Comment (Tailp) [reply dspot/src/main/java/eu/stamp_project/utils/program/InputConfiguration.java]: Hi @danglotb :) , I removed all collector related in InputConfig and instead added a new CollectorConfig class to be configured by Configuration.java and store information from JSAP . I can use this class to fetch information for each collector kind upon instatiation . So the base idea is the same , but this make the code a bit shorter since we skipped InputInformationCollector.

**Comment by monperrus** [comment]
Excellent!

--Martin
